### PR TITLE
Address multiple users & permissions issues

### PR DESCRIPTION
**Pull request by alexandrebodin**
### What does it do?

Addresses bug mentioned in https://github.com/strapi/strapi/pull/12322 and https://github.com/strapi/strapi/issues/12301

### Why is it needed?

See history in linked PR and issue

### How to test it?

- Try to register several users with same username or email and mixing them together with different providers

### Related issue(s)/PR(s)

- https://github.com/strapi/strapi/pull/12322 
- https://github.com/strapi/strapi/issues/12301)
- fix https://github.com/strapi/strapi/issues/13296
- fix https://github.com/strapi/strapi/issues/13217
- fix https://github.com/strapi/strapi/issues/12570
- fix https://github.com/strapi/strapi/issues/12362
- fix https://github.com/strapi/strapi/issues/12319
- fix https://github.com/strapi/strapi/issues/11806
- fix https://github.com/strapi/strapi/issues/12555
- fix https://github.com/strapi/strapi/issues/13172

### CODE REVIEW

**Comment by alexandrebodin** [comment]
@WalkingPizza If you can have a look here that would be awesome !

File: packages/plugins/users-permissions/server/services/providers.js [hunk 1]
```diff
@@ -43,6 +43,8 @@ module.exports = ({ strapi }) => {
   const connect = async (provider, query) => {
     const access_token = query.access_token || query.code || query.oauth_token;
 
+    console.dir({ query });
```
Comment (WalkingPizza) [inline packages/plugins/users-permissions/server/services/providers.js:46]: @alexandrebodin I think you forgot to remove this 😝

Comment (alexandrebodin) [reply packages/plugins/users-permissions/server/services/providers.js:46]: haha yes it's still a work in progress :D

**Comment by derrickmehaffy** [comment]
Testing the following based on f7f9ecc89868ec98debeb63b5c2f00f525fd34de

- https://github.com/strapi/strapi/pull/12322 
  - Confirmed, returns proper error
- https://github.com/strapi/strapi/issues/12301
  - Confirmed, can't register using a taken email as username :+1: 
- fix https://github.com/strapi/strapi/issues/13296
  - Confirmed, need to give the user's role access to `find` permission on roles but that's to be expected.
  - Also looks like by default we populate role in the `ctx.state.user` :+1: :partying_face: 
- fix https://github.com/strapi/strapi/issues/13217
  - Not tested but we were on Discord as you were working on this, trust it's fixed :stuck_out_tongue: 
- fix https://github.com/strapi/strapi/issues/12570
  - Fixed as the example changed
- fix https://github.com/strapi/strapi/issues/12362
  - Confirmed fixed, both for blocked/email conf we always check password first and return the proper error
- fix https://github.com/strapi/strapi/issues/12319
  - Confirmed fixed, both username and email must be unique
- fix https://github.com/strapi/strapi/issues/11806
  - Confirmed fixed

**Comment by derrickmehaffy** [comment]
Sorry @alexandrebodin didn't get a chance to test this before, I'll try testing tmrw when I'm at coworking space for the docs side

**Comment by WalkingPizza** [comment]
Haven't had the chance to test anything yet, but I noticed we are still relying on the Query Engine API for most operations in the U&P controllers.

Since quite a lot of things have been changed, should we also switch to the Entity Service API while we are at it?

**Comment by derrickmehaffy** [comment]
> Haven't had the chance to test anything yet, but I noticed we are still relying on the Query Engine API for most operations in the U&P controllers.
> 
> Since quite a lot of things have been changed, should we also switch to the Entity Service API while we are at it?

:sweat_smile: This is something I brought up internally and with @iicdii too and I think we decided to do as little modification as possible to just make sure stuff works without fundamentally rewriting/refactoring the plugin.

Soon (tm), and with a TBD date we plan to replace this plugin with a new one that will allow us to deprecate this one as we don't implement breaking changes that require us to bump this package to v5.x.x.

**Comment by alexandrebodin** [comment]
@WalkingPizza anywhere specific you think we would benefit from the entityService? I'll have a look

**Pull request merged**